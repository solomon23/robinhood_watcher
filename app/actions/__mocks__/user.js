const tryAuthFunction = jest.fn()

export function tryAuth() {
  return tryAuthFunction
}

export default {
  tryAuth,
}
