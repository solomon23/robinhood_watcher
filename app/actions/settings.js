// @flow
import { getSettings } from '../services/api'
import * as ipc from '../services/ipc'

export const VIEW_BY = {
  PERCENT: 'PERCENT',
  TOTAL: 'TOTAL',
  INDIVIDUAL: 'INDIVIDUAL',
}

export const SETTINGS_UPDATE = 'SETTINGS_UPDATE'
export function updateSettings(settings: UserSettings) {
  return { type: SETTINGS_UPDATE, payload: settings }
}

export function loadSettings() {
  return (dispatch: Dispatch) => {
    // get the settings from the local store
    const settings = getSettings()

    if (settings) {
      dispatch(updateSettings(settings))
    }
  }
}

export function saveSettings(settings: UserSettings) {
  return (dispatch: Dispatch) => {
    // update it locally
    dispatch(updateSettings(settings))

    // save it over the wire
    ipc.saveSettings(settings)
  }
}

export function quit() {
  return () => {
    ipc.quit()
  }
}

export function logout() {
  return () => {
    ipc.logout()
  }
}
