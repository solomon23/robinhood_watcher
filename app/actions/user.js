import { authenticate, tryAuthenticate } from '../services/api'
import type { Dispatch } from '../reducers/types'

export const LOGIN_USER_LOADING = 'LOGIN_USER_LOADING'
export const LOGIN_USER_ERROR = 'LOGIN_USER_ERROR'
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS'
export const LOGIN_USER_NEEDS_MFA = 'LOGIN_USER_NEEDS_MFA'
export const USER_LOGOUT = 'USER_LOGOUT'
export const SET_USER_ACCOUNT_NUMBER = 'SET_USER_ACCOUNT_NUMBER'

export function login(username, password, mfaCode) {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return authenticate(username, password, mfaCode)
      .then(response => {
        if (response.twoFactorAuthRequired && !mfaCode) {
          dispatch({ type: LOGIN_USER_NEEDS_MFA })
        } else if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function tryAuth() {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return tryAuthenticate()
      .then(response => {
        if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function logout() {
  return {
    type: USER_LOGOUT,
  }
}