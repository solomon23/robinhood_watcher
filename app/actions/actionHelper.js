// @flow
import * as userActions from './user'
import * as appActions from './app'

export function createRequestTypes(base: string) {
  return ['REQUEST', 'SUCCESS', 'FAILURE'].reduce((acc, type) => {
    acc[type] = `${base}_${type}`
    return acc
  }, {})
}

export function action(type: string, payload: any = {}) {
  return { type, ...payload }
}

export function createRequests(base: any) {
  return {
    request: (o: any) => action(base.REQUEST, o),
    success: (o: any) => action(base.SUCCESS, o),
    failure: (o: any) => action(base.FAILURE, o),
  }
}

export function doApiCall(options: any) {
  const defaults = {
    api: () => { },
    action: {},
    apiParams: {},
    guardFunc: () => true,
    postFunc: (dispatch, getState, response) => response,
  }
  const opts = { ...defaults, ...options }
  opts.action = Array.isArray(options.action) ? options.action : [options.action]

  return (dispatch: Dispatch, getState: GetState) => {
    // bail if we don't pass any rules sent in
    if (!opts.guardFunc(getState())) {
      return Promise.resolve()
    }

    const { user } = getState()
    if (!user.authenticated) {
      // bail
      return Promise.resolve()
    }

    const params = { ...opts.apiParams || {}, user }

    // mark the call as starting
    opts.action.map(act => dispatch(act.request({ request: params })))

    // call the api with the args
    return opts.api(params)
      .then(
        response => opts.postFunc(dispatch, getState, response)
      ).then((response) => {
        opts.action.map(act => dispatch(act.success({ request: params, response })))

        return response
      }, (error) => {
        opts.action.map(act => dispatch(act.failure({ request: params, response: error })))
        return Promise.reject(error)
      }).catch((error) => {
        if (error === 401) {
          // log the user out and stop refreshing the data since our auth token is no longer valid
          dispatch(userActions.logout())
          dispatch(appActions.stopRefresh())
        }
      })
  }
}
