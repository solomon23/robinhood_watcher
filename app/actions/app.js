// @flow
import type { GetState, Dispatch } from '../reducers/types'
import * as API from '../services/api'
import { createRequests, doApiCall, createRequestTypes } from './actionHelper'

export const START_REFRESH_TIMER = 'START_REFRESH_TIMER'
export const STOP_REFRESH_TIMER = 'STOP_REFRESH_TIMER'

export const GET_ACCOUNT_NUMBER = createRequestTypes('GET_ACCOUNT_NUMBER')
export const getAccountNumberActions = createRequests(GET_ACCOUNT_NUMBER)

export const GET_PORTFOLIO = createRequestTypes('GET_PORTFOLIO')
export const getPortfolioActions = createRequests(GET_PORTFOLIO)

export const GET_ACCOUNT_POSITIONS = createRequestTypes('GET_ACCOUNT_POSITIONS')
export const getAccountPositionsActions = createRequests(GET_ACCOUNT_POSITIONS)

export function startRefreshTimer() {
  return { type: START_REFRESH_TIMER }
}

export function stopRefresh() {
  return { type: STOP_REFRESH_TIMER }
}

export function getAccount() {
  return doApiCall({
    api: API.getAccountNumber,
    action: getAccountNumberActions,
    guardFunc: (state) => {
      const { user } = state
      return !user.accountNumber
    },
  })
}

export function getPortfolio() {
  return doApiCall({
    api: API.getPortfolio,
    action: getPortfolioActions,
  })
}

export function getAccountPositions() {
  return doApiCall({
    api: API.getAccountPositions,
    action: getAccountPositionsActions,
  })
}

export function getAllData() {
  return async function getAllCallback(dispatch: Dispatch, getState: GetState) {
    // (async () => {
      const { user } = getState()

      // bail if the user isn't logged in
      if (!user.authenticated) {
        return
      }

      if (!user.accountNumber) {
        await getAccount()(dispatch, getState)
      }

      // const positions = await API.getAccountPositions(user)
      await getPortfolio()(dispatch, getState)
      await getAccountPositions()(dispatch, getState)
    // })()
  }
}
