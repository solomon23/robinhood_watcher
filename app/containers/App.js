// @flow
import * as React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as UserActions from '../actions/user'
import Login from './Login'

function mapDispatchToProps(dispatch) {
  return bindActionCreators(UserActions, dispatch)
}

function mapStateToProps(state) {
  return { user: state.user }
}

type Props = {
  children: React.Node,
  user: object,
  tryAuth: () => void
}

class App extends React.Component<Props> {
  props: Props

  componentDidMount() {
    const { user, tryAuth } = this.props

    if (!user.authenticated) {
      // try to authenticate them
      tryAuth()
    }
  }

  render() {
    const { children, user } = this.props

    if (!user.authenticated) {
      return <React.Fragment><Login /></React.Fragment>
    }

    return <React.Fragment>{children}</React.Fragment>
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
