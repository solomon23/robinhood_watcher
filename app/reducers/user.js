// @flow
import * as userActions from '../actions/user'
import * as appActions from '../actions/app'
import type { Action } from './types'

const defaultState = {
  isLoading: false,
  isError: false,
  errorReason: null,
  needsMfa: false,
  authenticated: false,
  accountNumber: null,
}

export default function app(state = defaultState, action: Action) {
  switch (action.type) {
    case userActions.LOGIN_USER_LOADING:
      return { ...defaultState, isLoading: true, authenticated: false }

    case userActions.LOGIN_USER_ERROR:
      return { ...defaultState, isError: true, errorReason: action.message }

    case userActions.LOGIN_USER_NEEDS_MFA:
      return { ...defaultState, needsMfa: true, authenticated: false }

    case userActions.LOGIN_USER_SUCCESS:
      return { ...defaultState, authenticated: action.response.authenticated }

    case appActions.GET_ACCOUNT_NUMBER.SUCCESS:
      return { ...state, accountNumber: action.response }

    case userActions.USER_LOGOUT:
      return { ...defaultState, authenticated: false }

    default:
      return state
  }
}
