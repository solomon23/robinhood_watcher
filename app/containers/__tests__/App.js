
import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureMockStore from 'redux-mock-store'
import { shallowToJson } from 'enzyme-to-json'
import thunk from 'redux-thunk'
import App from '../App'
import { tryAuth } from '../../actions/user'

jest.mock('../../actions/user')

Enzyme.configure({ adapter: new Adapter() })
const mockStore = configureMockStore([thunk])

let store

describe('App component', () => {
  beforeEach(() => {
    store = { user: {} }
    tryAuth().mockClear()
  })

  it('should show login', () => {
    const wrapper = shallow(<App store={mockStore(store)} />).shallow()
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should fire try auth', () => {
    store.user.authenticated = false
    const mock = mockStore(store)
    shallow(<App store={mock} />).shallow()
    expect(tryAuth()).toHaveBeenCalled()
  })

  it('should show page', () => {
    store.user.authenticated = true
    const wrapper = shallow(<App store={mockStore(store)} />).shallow()

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
