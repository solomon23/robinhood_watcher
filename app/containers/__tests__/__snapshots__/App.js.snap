// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should show login 1`] = `
<React.Fragment>
  <Connect(Login) />
</React.Fragment>
`;

exports[`App component should show page 1`] = `<React.Fragment />`;
